src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -357,15 +357,22 @@ public override Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType m
 
             WebSocketValidate.ValidateArraySegment(buffer, nameof(buffer));
 
-            _writeBuffer ??= new MemoryStream();
-            _writeBuffer.Write(buffer.Array!, buffer.Offset, buffer.Count);
-
             if (!endOfMessage)
+            {
+                _writeBuffer ??= new MemoryStream();
+                _writeBuffer.Write(buffer.Array!, buffer.Offset, buffer.Count);
                 return Task.CompletedTask;
+            }
 
-            MemoryStream writtenBuffer = _writeBuffer;
+            MemoryStream? writtenBuffer = _writeBuffer;
             _writeBuffer = null;
 
+            if (writtenBuffer is not null)
+            {
+                writtenBuffer.Write(buffer.Array!, buffer.Offset, buffer.Count);
+            }
+            buffer = writtenBuffer?.ToArray() ?? buffer;
```

Comment:
Should be possible to store the writtenBuffer.Count and use TryGetBuffer here then fall back to ToArray ()

Reply:
Good call